pkg/core/combat/shape.go

```diff
@@ -1,131 +1,117 @@
 package combat
 
 import (
-	"fmt"
 	"math"
 )
 
 type Shape interface {
-	Positional
+	positional
 	IntersectCircle(c Circle) bool
 	IntersectRectangle(r Rectangle) bool
 	String() string
 }
 
-type Positional interface {
-	Pos() (x, y float64)
+type positional interface {
+	Pos() Point
 }
 
-func NewCircle(x, y, r float64) *Circle {
-	return &Circle{
-		x: x,
-		y: y,
-		r: r,
-	}
-}
-
-type SingleTarget struct {
-	Target TargetKey
+func DefaultDirection() Point {
+	return Point{X: 0, Y: 1}
 }
 
-func (s *SingleTarget) IntersectCircle(in Circle) bool       { return false }
-func (s *SingleTarget) IntersectRectangle(in Rectangle) bool { return false }
-func (s *SingleTarget) Pos() (float64, float64)              { return 0, 0 }
-func (s *SingleTarget) String() string                       { return fmt.Sprintf("single target: %v", s.Target) }
-
-// this is for attack that only hits self
-type SelfDamage struct{}
-
-func (c *SelfDamage) IntersectCircle(in Circle) bool       { return false }
-func (c *SelfDamage) IntersectRectangle(in Rectangle) bool { return false }
-
-type Circle struct {
-	x, y, r float64
+// dir needs to be magnitude of 1 if passing custom dir;
+// returns a new Point which is pos + rotated offset
+func CalcOffsetPoint(pos, offset, dir Point) Point {
+	if dir == DefaultDirection() {
+		return pos.Add(offset)
+	}
+	return pos.Add(offset.Rotate(dir))
 }
 
-func (c *Circle) String() string {
-	return fmt.Sprintf("r: %v x: %v y: %v", c.r, c.x, c.y)
+// https://wumbo.net/formulas/angle-between-two-vectors-2d/
+func CalcDirection(src, trg Point) Point {
+	// avoid division by 0 in Normalize()
+	if trg == src {
+		return DefaultDirection()
+	}
+	return trg.Sub(src).Normalize()
 }
 
-func (c *Circle) IntersectCircle(c2 Circle) bool {
-	//(R0 - R1)^2 <= (x0 - x1)^2 + (y0 - y1)^2 <= (R0 + R1)^2
-	return math.Pow(c.x-c2.x, 2)+math.Pow(c.y-c2.y, 2) < math.Pow(c.r+c2.r, 2)
+func AABBTest(a, b []Point) bool {
+	aMin := a[0]
+	aMax := a[1]
+	bMin := b[0]
+	bMax := b[1]
+	return aMin.X <= bMax.X && aMax.X >= bMin.X && aMin.Y <= bMax.Y && aMax.Y >= bMin.Y
 }
```

Comment:
Another place we need to think about the `==` case (or at least make sure it also does the `collide`/`true` on `==` if the more detailed tests to). Current implementation of AABB treats treats `==` as collision

Reply:
No change required since  `==` means `colliding`/`true`.